### ISSUE TITLE
Q2: screen Ouro task quality and establish a cached external baseline

### ISSUE THREAD

**Issue by hsliuustc0106** [ThinkFlowLab/vllm-rlt#8]
## Precision requirements — 2026-09-12

For new work, **BF16 is the primary GPU inference, kernel-validation, quality, and performance target**. Use BF16 weights, ordinary activations, and KV storage; specify accumulation independently and retain FP32 in numerically sensitive reductions/intermediates. Full-model FP32 is an optional diagnostic/reference configuration, not the default prerequisite. See the [precision policy](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/precision-policy.md).

BF16 profiling, matched optimization A/B, and quality data collection may proceed alongside the original Q1 reference diagnosis after each run's functional, finiteness, and ownership prerequisites pass. Report observations, reference fidelity, and qualification separately. Preserve exact cache/data-movement/RNG invariants and metadata-only A/B behavior; differing arithmetic paths use independently justified, predeclared BF16 numerical and decision/quality criteria. Universal bitwise or top-1 equality is not a blanket mixed-precision requirement, and near-tie exemptions cannot be invented after a run.

For dtype choices, follow explicit original Ouro paper guidance first, then the pinned release configuration and inference code. The [source audit](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/paper-notes.md#original-ouro-paper-and-precision-guidance) distinguishes published FP32 operations from backend/project choices; the paper has no per-kernel inference recipe.

This amendment supersedes conflicting dtype/eligibility requirements below for successor work. All dated contracts, raw results, original failures, and historical acceptance decisions remain unchanged. New criteria need a separately frozen contract; no benchmark or qualification is claimed by this edit.

Use fixed-depth BF16 (B16) as the primary quality baseline and adaptive BF16 (A16) as its paired comparison. F32 is an optional, separately justified sensitivity control. New cached external timing uses matched BF16 with an explicit generation/replay-equivalence contract. Historical FP32 quality/timing results and their limitations remain intact.

Parent roadmap: #2. Depends on M1 #3 for measurement/reporting and [Q1 #4](https://github.com/hsliuustc0106/vllm-lt/issues/4) for numerical qualification. BF16 quality and external-baseline data collection can proceed alongside the historical Q1 diagnosis. Numerical and quality qualification remain separate decisions.

## Goal and scope

Determine which numerical/depth configurations support a bounded quality statement, and establish one defensible comparison with the released Ouro implementation. These are separate experiments: a quality pass does not establish speed, and replay throughput does not establish quality.

The recorded FP32 smoke passed; BF16 exceeded its original logit tolerances despite matching three short continuations. BF16 enters the evaluation matrix under a predeclared mixed-precision contract; final qualification records the applicable Q1 evidence and any unresolved limitations. Q2 does not relax that failure or replace numerical acceptance with answer accuracy.

Non-goals: gate/threshold training or tuning, numerical fixes, kernel optimization, paper reproduction, broad benchmark rankings, serving, extra datasets/models, or an adaptive speed comparison against the official implementation's different KV semantics.

### Original Ouro accuracy protocol — 2026-09-12

The recorded quality screen is FP32 only: 4/64 correct, 58 unparseable answers, and 57 output-limit terminations. No paired BF16 task-accuracy result is established. [Source and protocol comparison](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/paper-notes.md#accuracy-evidence-and-the-original-evaluation-protocol).

The successor accuracy contract must follow original Ouro Appendix C.1: GSM8K, 3-shot chain-of-thought, strict match, and lm-eval-harness. Pin the harness/task revision, demonstrations, prompt format, answer extraction, EOS/stop rules, model revision, and token budgets before execution. Details the paper leaves unspecified remain declared project choices. Compare native and official fixed-four-loop BF16 on the same selected examples before assessing adaptive BF16.

This supersedes the old zero-shot prompt/parser and 512/256 limits below for new accuracy runs. The new plan must explicitly replace those fields, size budgets for few-shot prompts, and reconcile every native/official/adaptive configuration's record counts within a predeclared total budget. Preserve the original plan and 4/64 result. A 64-example screen is not a reproduction of the paper's full benchmark or its reported 78.92%.

- [ ] **AC-Q2-OURO — Author-aligned accuracy evidence.** Publish the new frozen protocol, paired native/official BF16 fixed-depth outcomes and uncertainty, output-limit/parser coverage, and separate adaptive results where selected. Record missing author settings explicitly; do not attribute a score gap to dtype without a controlled comparison.

## Quality contract

Freeze these inputs before any evaluation outputs are inspected:

- `ByteDance/Ouro-1.4B` and tokenizer revision `574fa66cb8bf5abdc979642d01cf2b79b16bfab1`; verify local content manifests, code SHA/modifications, environment, Q1 decision, arithmetic settings, exact reserved GPU IDs, and CPU/NUMA controls.
- GSM8K `main`/`test` from a recorded immutable dataset revision, including source IDs/content hashes. Format a fixed question-only prompt requesting reasoning followed by `#### <number>`. Commit the exact template and tokenizer options.
- Exclude prompts exceeding 512 tokens before selection; never truncate question text. Sort eligible source IDs by a documented seeded SHA-256 selection key; freeze the first 64 and two additional disjoint feasibility examples. Record the eligible population and exclusions. Claims concern this prompt-length-limited slice.
- Greedy decoding, natural EOS enabled, no forced output replay, maximum 256 generated tokens, one request at a time, full-depth prefill, Triton attention, and fixed cache/scheduler settings with verified capacity. The final sampled token requires no KV forward pass.

Run these native configurations, changing only the named factor within each comparison:

| Configuration | Policy | Comparison |
| --- | --- | --- |
| F32 | Optional FP32, fixed four loops | Separately justified dtype-sensitivity control. |
| B16 | BF16 with declared accumulation precision, fixed four loops | Primary quality baseline; optional F32 versus B16 isolates dtype. |
| A16 | Same BF16, threshold 0.7, minimum two/maximum four loops | B16 versus A16 evaluates the actual adaptive/last-exited policy. |

Each configuration generates its own complete token history. Do not teacher-force evaluation continuations, transplant exits, or remove examples because their outputs diverge. The first output always records depth four; exclude it from adaptive decode-depth averages.

Define a CPU-only answer parser before execution: use the final complete line matching `####` followed by a signed integer/decimal, normalize permitted thousands separators, and compare exact decimal values with the similarly extracted reference answer. Preserve parser version and fixtures covering malformed/absent markers, commas, signs, decimals, and multiple markers. Missing/unparseable answers count as incorrect. Record length-limited completions; score any valid final answer under the same rule. Timeouts/device failures make the experiment incomplete, not ordinary wrong answers.

## Paired quality decision

Before execution, require `max_loss_pp` for each planned comparison, a confidence level, and the interval method. No default loss margin is inferred from the observations. Report all 64 paired outcomes, accuracies, token lengths, truncation/parse failures, exit distributions, and disagreements.

Let `L` count baseline-correct/candidate-wrong cases and `G` count the reverse. Observed loss is `(L-G)/64`. Use a predeclared paired interval or, initially, a conservative one-sided bound: exact binomial upper bound for `L/64` minus the corresponding lower bound for `G/64`, splitting the error probability between the two bounds. Handle zero counts explicitly; a bootstrap producing a zero-width interval when every pair agrees is insufficient evidence of zero population loss. State per-comparison coverage; an overall joint claim needs an adjusted contract.

Qualify only when the upper loss bound is within the predeclared margin and all numerical/coverage gates pass. Otherwise record `inconclusive` or `failed` as appropriate. A 64-example screen can remain inconclusive even with no observed losses. Never tune threshold/template/parser on these examples or silently add samples. Broader claims require a separate budget and held-out evaluation.

## Equivalent external baseline

Prepare a locally inspected copy of the official model code at the same pinned checkpoint revision. Record code/dependency hashes and any compatibility adapter. Require cached incremental generation: each later call receives one newly generated input token, maintains distinct depth/layer KV history, and executes four loops. Verify this contract and fixed-depth outputs before timing; stop if the released interface cannot support it. The local dense oracle and full-prefix recomputation are not speed baselines.

For the successor comparison use native versus official **BF16 fixed-depth W1** with declared accumulation settings: one identical 128-token prompt, 64 outputs, greedy selection, `ignore_eos=True`, identical tokenizer IDs/arithmetic flags and full-depth history. Use M1 timing boundaries, including coda/selection/readback and the same observer policy; exclude loading/tokenization/compilation. In feasibility, verify finiteness, cache/history invariants, fixed depth, and equal prompt/output lengths and executed work. Assess actual output disagreements under the predeclared BF16 numerical/decision policy; they do not automatically invalidate matched-work timing. Identical-generation claims still require identical actual histories. Never silently substitute forced tokens or omit disagreements. Record attention/cache backend differences as implementation differences, including any inherent extra work. This result supports only its measured workload.

## Interfaces and artifacts

Reuse M1's planning/manifests/results; add small CPU preparation/scoring helpers and a private external adapter, not another harness framework. Proposed commands are `prepare-quality`, `run-quality`, `score-quality`, and `run-external`, each accepting a frozen plan and fresh output directory.

All JSON artifacts use `schema_version: 1` and a type discriminator:

- `QualityPlan`: dataset/selection/template/parser hashes, exact question IDs/token IDs, expected-answer provenance, configuration IDs, Q1 dependency hashes, margins/interval policy, budgets and controls.
- `QualityRecord`: example/configuration IDs, prompt hash, complete output IDs/text, exit depths, finish reason, parsed/reference values, correctness, truncation, and execution/failure status.
- `QualityReport`: denominator/coverage, paired contingency counts, loss estimates/bounds, qualification decisions and reasons. Scoring requires no GPU or weights.
- `ExternalPlan/Result`: pinned implementation manifests, adapter/cache equivalence checks, W1 fixture, M1 timing records, raw paired values/ranges, and comparison eligibility.

The external adapter serves one request at a time in this initial comparison:

```python
class CachedOuroAdapter:
    def start(self, prompt_ids: Sequence[int], *, controls: ExternalControls) -> TokenPrediction: ...
    def advance(self, input_token_id: int, *, position: int) -> TokenPrediction: ...
    def reset_request(self) -> None: ...
    def close(self) -> None: ...
```

`start` owns prompt prefill and the first coda/greedy result. `advance` consumes exactly one preceding generated token, retains the adapter's depth/layer cache, and includes coda/greedy/readback. `TokenPrediction` contains the actual token ID, output index and full exit depth four; the shared harness assigns delivery timestamps on return. The adapter exposes no forced-token path for this comparison. Validate sequential positions, reset after completion, and release request KV while retaining prepared model weights. Implementation-specific cache APIs stay inside the adapter; sampler/timing semantics remain in the frozen contract.

The CPU scorer interface is `score_records(plan, records) -> QualityReport`: require exactly the declared example/configuration keys, reject duplicates, retain invalid/timeout records as missing execution coverage, and compute parser/statistical results without modifying source generations. A parser or margin revision creates a new identified scoring contract and preserves the original report.

## Run budget and failure handling

CPU probes validate manifests, parser/statistical edge cases, capacity, and adapter/controller contracts without full weights or accelerator access.

Quality: one excluded feasibility pass on the two disjoint examples per eligible configuration; then **one deterministic pass per configuration**, explicitly overriding repeated performance runs. No additional quality warmups or repeat samples. Allow a two-hour reservation and ten minutes per example, with a fixture-derived step bound; the overall cap wins.

External timing: two cells, one excluded feasibility and one warmup per cell, then four measured executions ordered native–official–official–native. Retain prepared weights with a preflighted residency policy; reset request/KV state outside timers and preserve compiled/allocator caches under M1's policy. Declare a one-hour overall cap and ten minutes per execution. Use the same exact reserved devices and controls within comparisons. No profiler captures are added here.

Separate downloads/preparation/warmups from measured execution. Reuse environments and preserve shared caches. Persist started/completed records incrementally, abort only task-owned requests, and stop device work after CUDA/cleanup failure. Missing examples/runs produce incomplete coverage, never a silently reduced denominator or replacement trials.

## Acceptance criteria / 验收标准

- [ ] **AC-Q2-01 — Freeze prerequisites and decisions.** Before evaluation, preserve the dataset/model/tokenizer/source manifests, the exact 64 selected examples plus two disjoint feasibility examples, prompt/parser hashes, configuration controls, and applicable Q1 #4 numerical decision. Each paired comparison declares `max_loss_pp`, confidence level, interval method, and failure/qualification rules before outputs are inspected. B16/A16 collection requires the new run's functional/finiteness/ownership prerequisites, not closure of the historical Q1 diagnosis. Final qualification includes the applicable numerical and quality gates; unresolved reference fidelity stays explicit. No retrospective margin, parser, threshold, or sample tuning qualifies this experiment.

- [ ] **AC-Q2-02 — Complete the quality matrix.** Require exactly 64 unique completed evaluation records per declared configuration: **128 records for primary B16/A16, or 192 if the optional F32 control is predeclared**, excluding feasibility. Every record preserves actual greedy output IDs/text, exit depths, finish reason, parsed/reference answers, and correctness. Validate first-output depth four, fixed-policy depths four, and A16's declared actual gate/depth policy; exclude first outputs from decode-depth means. The same selected examples remain paired across configurations; natural EOS and the 256-token cap apply uniformly. Invalid answers count as incorrect, length-limited outputs retain their declared scoring rule, and timeouts/missing generations count as incomplete coverage. No forced continuations, replacement examples, dropped disagreements, or reduced denominators are permitted.

- [ ] **AC-Q2-03 — Verify scoring and uncertainty independently.** CPU tests cover parser edge cases, duplicate/missing record rejection, loss-sign/unit conversion, all paired contingency cells, and exact-bound endpoints including zero disagreements. Offline scoring reproduces all answers, contingency counts, accuracies, and bounds from raw records. For each comparison report `L`, `G`, observed loss `100*(L-G)/64`, and the predeclared upper loss bound in percentage points. Qualification requires that bound to be within `max_loss_pp` plus all prerequisite/coverage gates. Otherwise apply only the frozen failure rule or report statistical inconclusiveness; agreement on 64 cases cannot justify a zero-uncertainty population claim.

- [ ] **AC-Q2-04 — Prove external comparability before timing.** Link reviewed pinned official code/adapter evidence demonstrating one full-depth prefill followed by 63 one-token cached advances for W1's 64 outputs, separate depth/layer KV histories, correct positions, actual greedy selection/readback, and release/reset. Native and official feasibility executions must use the same 128-token input, produce 64 actual outputs at depth four, and pass the declared BF16/accumulation, functional, and workload-comparability checks. Record output disagreements and their numerical/decision status. Full-prefix recomputation, hidden forced outputs, shared-depth KV, or unequal work invalidate this cached matched-work comparison; identical-generation claims additionally require identical actual IDs.

- [ ] **AC-Q2-05 — Complete the bounded timing comparison.** Preserve two excluded feasibility executions, two warmups, and **four valid measured executions** ordered native–official–official–native: two complete matched pairs. Record exact device/control hashes, residency/reset policy, M1 clock/observer semantics, raw observations and ranges, and declared implementation differences. Every measured run completes all outputs and request cleanup. Missing runs, failed equivalence, incompatible controls, or budget exhaustion leave this deliverable open. A slower native result is valid evidence; no speedup is required.

- [ ] **AC-Q2-06 — Publish an auditable bounded report.** Offline reports reproduce metrics/decisions and reconcile all expected/completed/missing records. Preserve commands, manifests, raw outputs/timing records, parser/interval versions, cleanup evidence, exclusions, and workload limitations. Respect the existing one-pass quality, two-hour quality, four-run external, and one-hour external budgets; no automatic repeats, extra samples, or captures.

**Closure rule:** AC-Q2-01 through AC-Q2-06 must pass with both deliverables complete. A statistically inconclusive or negative result from a fully covered quality screen may close this bounded evaluation; missing execution coverage cannot. An unresolved numerical decision remains a qualification limitation and does not erase complete diagnostic quality evidence. Record qualification separately from issue completion: BF16/adaptive policies remain experimental unless their numerical and quality gates pass. The external result supports only the measured fixed-depth W1 comparison, never adaptive quality or general serving claims.

Reference boundary: [pinned official Ouro code](https://huggingface.co/ByteDance/Ouro-1.4B/blob/574fa66cb8bf5abdc979642d01cf2b79b16bfab1/modeling_ouro.py) and [recorded validation/comparability limits](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/docs/paper-notes.md).
